### PR TITLE
Feat: Added support for Standing Instructions

### PR DESCRIPTION
**Pull request by devansh-299**
## Description
Added complete support for Standing Instructions. Users can now create new standing instructions, get client specific standing instructions and update/delete standing instructions created earlier.
**Note:** Currently the update and create functionality are not working because of Fineract issues

### Creating New Standing Instruction
GIF-200611_213537
---

### Existing Standing Instruction
GIF-200611_214251

---

## Screen Shots 
|||||

---

- [x] Apply the `AndroidStyle.xml` style template to your code in Android Studio.

- [x] Run the unit tests with `./gradlew check` to make sure you didn't break anything

- [x] If you have multiple commits please combine them into one commit by squashing them.

### CODE REVIEW

File: core/src/main/java/org/mifos/mobilewallet/core/data/fineract/entity/payload/StandingInstructionPayload.kt [hunk 1]
```diff
@@ -0,0 +1,85 @@
+package org.mifos.mobilewallet.core.data.fineract.entity.payload
+
+import android.os.Parcel
+import android.os.Parcelable
+
```
Comment (jawidMuhammadi) [inline core/src/main/java/org/mifos/mobilewallet/core/data/fineract/entity/payload/StandingInstructionPayload.kt:5]: Instead of overriding all these methods from `Parcelable`, you can simply Add `@Parcelize` annotation here (on top of `data class  StandingInstructionPayload`) then you can simply delete all overridden methods. This [article](https://medium.com/@BladeCoder/a-study-of-the-parcelize-feature-from-kotlin-android-extensions-59a5adcd5909) might help if you want to learn how it works.

**Review by jawidMuhammadi — CHANGES_REQUESTED** [review]

File: mifospay/src/main/res/layout/activity_si_details.xml [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+                        android:textColor="@color/colorAccent"
+                        android:textSize="@dimen/value_16sp"
+                        android:layout_marginBottom="@dimen/marginItemsInSectionMedium" />
+
+                    <TextView
+                        android:layout_width="wrap_content"
+                        android:layout_height="wrap_content"
+                        android:textColor="@color/gray_dark"
+                        android:textSize="@dimen/value_16sp"
+                        android:text="Amount:"/>
```
Comment (jawidMuhammadi) [inline mifospay/src/main/res/layout/activity_si_details.xml]: It would be much better if it wasn't hardcoded otherwise will not be able to translate it :)

File: mifospay/src/main/res/layout/activity_new_si.xml [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+                android:layout_height="@dimen/value_50dp"
+                android:text="SELECT DATE"
+                android:layout_gravity="center"
+                android:layout_marginBottom="@dimen/value_20dp"
+                android:textColor="@color/white" />
+            <Button
+                android:id="@+id/btn_create_si"
+                android:layout_width="260dp"
+                android:layout_height="@dimen/value_50dp"
+                android:text="Submit"
```
Comment (jawidMuhammadi) [inline mifospay/src/main/res/layout/activity_new_si.xml]: same here

File: mifospay/src/main/java/org/mifos/mobilewallet/mifospay/standinginstruction/ui/SIDetailsActivity.kt [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+
+        val res = resources
+        iv_empty_no_transaction_history
+                    .setImageDrawable(res.getDrawable(drawable))
+        tv_empty_no_transaction_history_title.text = res.getString(errorTitle)
+        tv_empty_no_transaction_history_subtitle.text = res.getString(errorMessage)
+    }
+
+    override fun siDeletedSuccessfully() {
+        showToast("Deleted Successfully")
```
Comment (jawidMuhammadi) [inline mifospay/src/main/java/org/mifos/mobilewallet/mifospay/standinginstruction/ui/SIDetailsActivity.kt]: Same here

File: mifospay/src/main/java/org/mifos/mobilewallet/mifospay/standinginstruction/ui/SIDetailsActivity.kt [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+        tv_si_from_name.text = "Name: ${standingInstruction.fromClient.displayName}"
+        tv_si_from_number.text = "Number: ${standingInstruction.fromAccount.accountNo}"
+        tv_si_to_name.text = "Name: ${standingInstruction.toClient.displayName}"
+        tv_si_to_number.text = "Number: ${standingInstruction.toAccount.accountNo}"
+
+        tv_si_status.text = standingInstruction.status.value
+        tv_recurrence_interval.text = standingInstruction.recurrenceInterval.toString()
+
+        // setting up TextInputLayouts
+        til_si_edit_amount.hint = "₹ ${standingInstruction.amount}"
```
Comment (jawidMuhammadi) [inline mifospay/src/main/java/org/mifos/mobilewallet/mifospay/standinginstruction/ui/SIDetailsActivity.kt]: Are we sure this amount will be in Rupee all the time? I think the symbol should be changed based on amount's currency dynamically.

File: mifospay/src/main/java/org/mifos/mobilewallet/mifospay/standinginstruction/ui/SIDetailsActivity.kt [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+
+        tv_valid_from.text = "${standingInstruction.validFrom[2]}-" +
+                "${standingInstruction.validFrom[1]}-${standingInstruction.validFrom[0]}"
+        tv_valid_till.text = "${standingInstruction.validTill?.get(2)}-" +
+                "${standingInstruction.validTill?.get(1)}-${standingInstruction.validTill?.get(0)}"
+
+        tv_si_from_name.text = "Name: ${standingInstruction.fromClient.displayName}"
+        tv_si_from_number.text = "Number: ${standingInstruction.fromAccount.accountNo}"
+        tv_si_to_name.text = "Name: ${standingInstruction.toClient.displayName}"
+        tv_si_to_number.text = "Number: ${standingInstruction.toAccount.accountNo}"
```
Comment (jawidMuhammadi) [inline mifospay/src/main/java/org/mifos/mobilewallet/mifospay/standinginstruction/ui/SIDetailsActivity.kt]: Can we use `formatted string` dynamically in here? Like `<string name="name">Name: %1$s</string>`

**Review by jawidMuhammadi — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/mifos/mobilewallet/core/domain/usecase/standinginstruction/CreateStandingTransaction.kt [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+                                if (account.productId ==
+                                        Constants.WALLET_ACCOUNT_SAVINGS_PRODUCT_ID) {
+                                    walletAccount = account
+                                    break
+                                }
+                            }
+                            walletAccount?.let{
+                                toAccount = walletAccount
+                                createNewStandingInstruction()
+                            }?: useCaseCallback.onError(Constants.NO_WALLET_FOUND)
```
Comment (jawidMuhammadi) [inline core/src/main/java/org/mifos/mobilewallet/core/domain/usecase/standinginstruction/CreateStandingTransaction.kt:127]: `Constants.NO_WALLET_FOUND` do we show this error message in UI? If yes then it would be much better to add inside `string.xml` file

File: mifospay/src/main/java/org/mifos/mobilewallet/mifospay/standinginstruction/adapter/StandingInstructionAdapter.kt [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
+
+        fun bindItems(standingInstruction: StandingInstruction) {
+            val tvFromClientName = itemView.findViewById(R.id.tv_from_client_name) as TextView
+            val tvToClientName = itemView.findViewById(R.id.tv_to_client_name) as TextView
+            val tvValidTill  = itemView.findViewById(R.id.tv_valid_till_date) as TextView
+            val tvAmount  = itemView.findViewById(R.id.tv_amount) as TextView
+
+            tvFromClientName.text = "From: ${standingInstruction.fromClient.displayName}"
```
Comment (jawidMuhammadi) [inline mifospay/src/main/java/org/mifos/mobilewallet/mifospay/standinginstruction/adapter/StandingInstructionAdapter.kt]: Use formatted string, please.

File: mifospay/src/main/java/org/mifos/mobilewallet/mifospay/standinginstruction/adapter/StandingInstructionAdapter.kt [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+            tvToClientName.text = "To: ${standingInstruction.toClient.displayName}"
+            /**
+             * Using hardcoded Currency as response doesn't return the currency
+             */
+            tvAmount.text = "${Constants.RUPEE} ${standingInstruction.amount.toString()}"
+            val validTill  = " ${standingInstruction.validTill?.get(2)}" +
+                    "-${standingInstruction.validTill?.get(1)}" +
+                    "-${standingInstruction.validTill?.get(0)}"
+            tvValidTill.text = "Valid Till: $validTill"
```
Comment (jawidMuhammadi) [inline mifospay/src/main/java/org/mifos/mobilewallet/mifospay/standinginstruction/adapter/StandingInstructionAdapter.kt]: same here

**Review by jawidMuhammadi — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/mifos/mobilewallet/core/domain/usecase/standinginstruction/CreateStandingTransaction.kt [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+
+                        } else {
+                            useCaseCallback.onError(Constants.ERROR_FETCHING_TO_ACCOUNT)
+                        }
+                    }
+                })
+    }
+
+    private fun createNewStandingInstruction() {
```
Comment (s-ayush2903) [inline core/src/main/java/org/mifos/mobilewallet/core/domain/usecase/standinginstruction/CreateStandingTransaction.kt:136]: @devansh-299 It would be much better if we use factory for initializing/creating new standingInstructions throughout the codebase.
Why?
* As it would reduce the following kind/type of repititive code in the concerned file(s) where it is needed to create a new standingInstruct'n

Comment (devansh-299) [reply core/src/main/java/org/mifos/mobilewallet/core/domain/usecase/standinginstruction/CreateStandingTransaction.kt:136]: @s-ayush2903 I get what that factory pattern will reduce the boilerplate code but we just need this code at two places, hence creating a factory for this in my opinion will be an overkill

File: core/src/main/java/org/mifos/mobilewallet/core/domain/usecase/standinginstruction/CreateStandingTransaction.kt (hunk 7 above)
Comment (devansh-299) [reply core/src/main/java/org/mifos/mobilewallet/core/domain/usecase/standinginstruction/CreateStandingTransaction.kt:127]: Thanks for pointing this out

File: mifospay/src/main/java/org/mifos/mobilewallet/mifospay/standinginstruction/ui/SIDetailsActivity.kt (hunk 5 above)
Comment (devansh-299) [reply mifospay/src/main/java/org/mifos/mobilewallet/mifospay/standinginstruction/ui/SIDetailsActivity.kt]: I forgot to mention in comments here, but the response doesn't specify the currency so for the time being, we are going with the hardcoded currency

File: core/src/main/java/org/mifos/mobilewallet/core/domain/usecase/standinginstruction/CreateStandingTransaction.kt (hunk 10 above)
Comment (s-ayush2903) [reply core/src/main/java/org/mifos/mobilewallet/core/domain/usecase/standinginstruction/CreateStandingTransaction.kt:136]: I'm astonished at consideration of factory as an _overkill_.Let me provide a few more instances which clicked right now(while authoring this thread)where this can be helpful
* Consider the scenario where one is trying to test this feature via api calls first;
* Then as for instruction creation takes 3 mandatory parameters, out of which vpa needs to be validated, so for end-to-end testing(covering all edge cases) one'll have to create two `SIs`-
a valid one and an invalid one,
* If we consider the unit test it solely(not networking calls, just its functionality), that'll also require same two more instances of the same
* So considering this minimum applicability/usages of SIs, the total locs would be=
  `41x(2+2+2)=246` {41=179-136} 
If we had factory then, it would be:- `41x1 + 4+4+4 = 53` {4=2x(1+1)}

But still, i was just putting forward the standard software design pattern and my invaluable view regarding the same, everything is upto you

Comment (devansh-299) [reply core/src/main/java/org/mifos/mobilewallet/core/domain/usecase/standinginstruction/CreateStandingTransaction.kt:136]: @s-ayush2903 thank you providing such an in depth analysis and for staging your views. I will consult my mentors over this but anyways I appreciate your thoughtfulness

**Review by shiv07tiwari — APPROVED** [review]
